src/components/withNavigationFallback.tsx

```diff
@@ -0,0 +1,43 @@
+import React, {ComponentType, RefAttributes, ForwardedRef, ReactElement, forwardRef, useContext, useMemo} from 'react';
+import {NavigationContext} from '@react-navigation/core';
+import {NavigationProp} from '@react-navigation/native';
+import {ParamListBase} from '@react-navigation/routers';
+import getComponentDisplayName from '../libs/getComponentDisplayName';
+
+type AddListenerCallback = () => void;
+
+type RemoveListenerCallback = () => void;
+
+type NavigationContextValue = {
+    isFocused: () => boolean;
+    addListener: () => AddListenerCallback;
+    removeListener: () => RemoveListenerCallback;
+};
+
+export default function <TProps, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>): (props: TProps & RefAttributes<TRef>) => ReactElement | null {
+    function WithNavigationFallback(props: TProps, ref: ForwardedRef<TRef>) {
+        const context = useContext(NavigationContext);
+
+        const navigationContextValue: NavigationContextValue = useMemo(() => ({isFocused: () => true, addListener: () => () => {}, removeListener: () => () => {}}), []);
+
+        return context ? (
+            <WrappedComponent
+                // eslint-disable-next-line react/jsx-props-no-spreading
+                {...props}
+                ref={ref}
+            />
+        ) : (
+            <NavigationContext.Provider value={navigationContextValue as unknown as NavigationProp<ParamListBase>}>
```

Comment:
Why do we need this assertion?

Reply:
@fabioh8010 Because `NavigationContext.Provider` value is expected to be of type `NavigationProp<ParamListBase>`, but to make our navigationContextValue be of this type, we need to add a lot of methods like navigate, reset, goBack, etc. If you have any ideas on how to improve it, please let me know.

Reply:
I'm trying to understand why do we pass only this values in the first place 🤔. It seems that only `Button` component uses this HOC, could you check what props from this HOC are really being injected in the component?

Reply:
@fabioh8010 Let me provide you with a little bit more context.
This HOC is a workaround for cases when Button/Composer is used outside of the Navigation container. `isFocused` param is used in button,  `addListener` and `removeListener` are used in Composer.
It's the second PR for this file to migrate, [that was the first](https://github.com/Expensify/App/pull/29449).
In the previous one, [we had a suggestion](https://expensify.slack.com/archives/C01GTK53T8Q/p1697205043249289) that this HOC isn't necessary anymore and removed it. It caused [regressions](https://github.com/Expensify/App/issues/29839).
This time I investigated it more careful, and it's not possible to get rid of this HOC, because ErrorBoundary is [located outside of the NavigationContainer](https://github.com/Expensify/App/blob/main/src/App.js#L72) an has a fallback component GenericErrorPage [which has Buttons inside](https://github.com/Expensify/App/blob/main/src/pages/ErrorPage/GenericErrorPage.js#L58). So without this HOC usage it'll crash.

For more details, see
https://github.com/Expensify/App/issues/24953#issuecomment-1771241283
https://github.com/Expensify/App/issues/24953#issuecomment-1772589273

Reply:
Ok I see, I guess we have no other option then